### DISCUSSION TITLE
V3 Rule attributes with Password Rules

### DISCUSSION THREAD

**Discussion by devhoussam1998** [livewire/livewire#5811] [1. Help]
I get this error message while using

**Constant expression contains invalid operations**

```
    #[Rule([
        'required', 
        Rules\Password::min(8)
        ->letters()
        ->mixedCase()
        ->numbers()
        ->symbols()
        ->uncompromised(3)
    ])]
    public $password = '';
```

**Comment by joshhanley** [comment]
@devhoussam1998 I don't think that will work because you can't instantiate an object in an attribute, which `Password:min()` is doing. I think you will still need to use the rules array for that.

**Reply by nuernbergerA** [reply]
we could support closures

**Reply by joshhanley** [reply]
@nuernbergerA that works in attributes? If so then that'd be good!

**Reply by nuernbergerA** [reply]
@joshhanley unfortunately only scalar types and ::class is supported :disappointed: 

its the same as a class [const](https://github.com/php/php-src/blob/9122638ecd7dfee1cbd141a15a8d59bfc47f6ab3/Zend/zend_compile.c#L8500-L8514)

![image](https://github.com/livewire/livewire/assets/13331388/e58f1222-a85c-4ff8-8cfc-a43f6581514b)
[source](https://stitcher.io/blog/attributes-in-php-8)

**Reply by joshhanley** [reply]
@nuernbergerA bugger! 😕 I think it may need a note in the docs about using attributes and instantiated rule objects don't work.

**Comment by uadoronin** [comment]
The same problem, how to add attribute names from lang to rule?
`#[Rule('required', attribute: __('page.field.email'))]`

Livewire v2 had property:
`protected $validationAttributes = [
        'email' => 'email address'
    ];`

What about in v3? Maybe:
`$this->validate(
         rules: [],
         messages: [],
         attributes: [
            'email' => __('page.field.email')
]);`

**Reply by joshhanley** [reply]
@uadoronin yeah same issue as above `__(...)` is a function call, which can't be passed into an attribute. For for now I would use the V2 style using `$rules/$validationAttributes` properties.

**Reply by flying-elephant-git** [reply]
I've solved this by creating a custom attribute that extends the rule attribute:

`
namespace App\Livewire\Attributes;

use Livewire\Attributes\Rule;

#[\Attribute]
class LocalizedRule extends Rule
{
    public function __construct(
        public $rule,
        protected $attribute = null,
        protected $as = null,
        protected $message = null,
        protected $onUpdate = true,
    ) {
        $this->as = $this->as ? __($this->as) : null;
        $this->message = $this->message ? __($this->message) : null;
    }
}

`

**Reply by viliusvsx** [reply]
Nice, I think this could work. But the main reason why this is a bad idea is that IDE won't recognize translations in the file like `validation.php`. Every time you will have to check that translation and that could be a huge pain.
